Admin/BreadcrumbsBuilder.php

```diff
@@ -0,0 +1,143 @@
+        }
+
+        $menu = $admin->buildBreadcrumbs($action);
+
+        do {
+            $breadcrumbs[] = $menu;
+        } while ($menu = $menu->getParent());
+
+        $breadcrumbs = array_reverse($breadcrumbs);
+        array_shift($breadcrumbs);
+
+        return $breadcrumbs;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildBreadcrumbs(
+        AdminInterface $admin,
+        $action,
+        MenuItemInterface $menu = null
+    ) {
+        if (!$menu) {
+            $menu = $admin->getMenuFactory()->createItem('root');
+
+            $menu = $this->createMenuItem(
+                $admin,
+                $menu,
+                'dashboard',
+                'SonataAdminBundle',
+                array('uri' => $admin->getRouteGenerator()->generate(
+                    'sonata_admin_dashboard'
+                ))
+            );
+        }
+
+        $menu = $this->createMenuItem(
+            $admin,
+            $menu,
+            sprintf('%s_list', $admin->getClassnameLabel()),
+            null,
+            array(
+                'uri' => $admin->hasRoute('list') && $admin->isGranted('LIST') ?
+                $admin->generateUrl('list') :
+                null,
+            )
+        );
+
+        $childAdmin = $admin->getCurrentChildAdmin();
+
+        if ($childAdmin) {
+            $id = $admin->getRequest()->get($admin->getIdParameter());
+
+            $menu = $menu->addChild(
+                $admin->toString($admin->getSubject()),
+                array(
+                    'uri' => $admin->hasRoute('edit') && $admin->isGranted('EDIT') ?
+                    $admin->generateUrl('edit', array('id' => $id)) :
+                    null,
+                )
+            );
+
+            return $childAdmin->buildBreadcrumbs($action, $menu);
+        }
+
+        if ('list' === $action && $admin->isChild()) {
+            $menu->setUri(false);
+        } elseif ('create' !== $action && $admin->hasSubject()) {
+            $menu = $menu->addChild($admin->toString($admin->getSubject()));
+        } else {
+            $menu = $this->createMenuItem(
+                $admin,
+                $menu,
+                sprintf('%s_%s', $admin->getClassnameLabel(), $action)
+            );
+        }
+
+        return $menu;
+    }
+
+    /**
+     * Creates a new menu item from a simple name. The name is normalized and
+     * translated with the specified translation domain.
+     *
+     * @param AdminInterface    $admin             used for translation
+     * @param MenuItemInterface $menu              will be modified and returned
+     * @param string            $name              the source of the final label
+     * @param string            $translationDomain for label translation
+     * @param array             $options           menu item options
+     *
+     * @return MenuItemInterface
+     */
+    private function createMenuItem(
```

Comment:
Plus, the Magento example is a counter-argument : refactoring it to something better would take weeks, but making it multiline is the very first step to making it more maintainable. I think seeing a multiline signature can draw attention on the fact that the signature might be too big.